UnitySDK/Assets/ML-Agents/Scripts/Sensor/RayPerceptionSensor.cs

```diff
@@ -0,0 +1,202 @@
+        /// <param name="detectableObjects">List of tags which correspond to object types agent can see</param>
+        /// <param name="startOffset">Starting height offset of ray from center of agent.</param>
+        /// <param name="endOffset">Ending height offset of ray from center of agent.</param>
+        /// <param name="castRadius">Radius of the sphere to use for spherecasting. If 0 or less, rays are used
+        /// instead - this may be faster, especially for complex environments.</param>
+        /// <param name="transform">Transform of the GameObject</param>
+        /// <param name="perceptionBuffer">Output array of floats. Must be (num rays) * (num tags + 2) in size.</param>
+        /// <param name="legacyHitFractionBehavior">Whether to use the legacy behavior for hit fractions.</param>
+        public static void PerceiveStatic(float rayLength,
+            IReadOnlyList<float> rayAngles, IReadOnlyList<string> detectableObjects,
+            float startOffset, float endOffset, float castRadius,
+            Transform transform, float[] perceptionBuffer,
+            bool legacyHitFractionBehavior = false)
+        {
+            Array.Clear(perceptionBuffer, 0, perceptionBuffer.Length);
+
+            // For each ray sublist stores categorical information on detected object
+            // along with object distance.
+            int bufferOffset = 0;
+            foreach (var angle in rayAngles)
+            {
+                Vector3 startPositionLocal = new Vector3(0, startOffset, 0);
+                Vector3 endPositionLocal = PolarToCartesian(rayLength, angle);
+                endPositionLocal.y += endOffset;
+
+                var startPositionWorld = transform.TransformPoint(startPositionLocal);
+                var endPositionWorld = transform.TransformPoint(endPositionLocal);
+
+                var rayDirection = endPositionWorld - startPositionWorld;
+                if (Application.isEditor)
+                {
+                    Debug.DrawRay(startPositionWorld,rayDirection, Color.black, 0.01f, true);
+                }
+
+                // Do the cast and assign the hit information for each detectable object.
+                //     sublist[0           ] <- did hit detectableObjects[0]
+                //     ...
+                //     sublist[numObjects-1] <- did hit detectableObjects[numObjects-1]
+                //     sublist[numObjects  ] <- 1 if missed else 0
+                //     sublist[numObjects+1] <- hit fraction (or 1 if no hit)
+                // The legacyHitFractionBehavior changes the behavior to be backwards compatible but has some
+                // counter-intuitive behavior:
+                //  * if the cast hits a object that's not in the detectableObjects list, all results are 0
+                //  * if the cast doesn't hit, the hit fraction field is 0
+
+                bool castHit;
+                RaycastHit rayHit;
+                if (castRadius > 0f)
```

Comment:
The size of the spherecast is now adjustable. If set to 0 (or smaller) it will use raycasts instead.